### PR TITLE
Make development setup easier with one-line make command

### PR DESCRIPTION
**Pull request by Alon-Alexander**
Hi, this is my first PR to `vaex`.
I would really like to start contributing to the project, and starting out I had a rough time installing `vaex` for development.
I decided to try and make it easier to install for development as my first issue.

Eventually I also made `vaex-core`'s `setup.py` a bit more standard (using `Pybind11Extension`).

_Note_: I didn't fully understand the `vendor` folder. Not its purpose nor how it is supposed to be used.
Thus I might have changed things there not in the way it was designed to be, so I would like an explanation on how to use it in the designed way.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -13,8 +13,32 @@ coverage:
...
+	rm -rf smæll2.parquet smæll2.yaml tests/data/parquet/ tests/data/smæll2.csv.hdf5 tests/data/unittest.parquet
+	find . -wholename "./tests/data/parquet_dataset_partitioned_*" -delete
+
+init:
+	make deps
+	python -m pip install -e .[dev]
+
+uninstall:
+	 @pip uninstall -y `pip list | grep -o -E "^vaex([-a-zA-Z0-9]+)?"` 2> /dev/null ||\
+	 	echo "Can't uninstall: Vaex is not installed"
```
Comment (maartenbreddels) [inline Makefile:36]: love it!

**Review by maartenbreddels — CHANGES_REQUESTED** [review]
Great start, love that you first start attacking issues you encounter.

The vendor directory is where the git submodules live, and as documented at `https://vaex.io/docs/installing.html#for-developers` you'll get them if you run:
`git clone --recursive https://github.com/vaexio/vaex`.

As everyone (including me) simply does a `git clone <URL>` you'll miss those, and get a very vague error msg when running setup.py. Maybe a good fix for that would be to check the directories (e.g. `packages/vaex-core/vendor/hopscotch-map`) in setup.py and if they are empty, suggest the user to run `git submodule update --init` which should fix that.

What do you think?

cheers,

Maarten

File: packages/vaex-core/src/vaexfast.cpp [hunk 2]
```diff
@@ -2122,7 +2122,7 @@ static struct PyModuleDef moduledef = {
 #define INITERROR return
 #endif
 
-
+__attribute__((visibility("default")))
```
[the remark this answers is not in the document]
Comment (Alon-Alexander) [reply packages/vaex-core/src/vaexfast.cpp:2125]: When changing to `pybind11`'s extension, it automatically adds some default (recommended) configuration for the compilation. One of the options is (on non-windows) `-fvisibility=hidden`, which makes all symbols (like functions) local by default, which means they are not exported.
That helps to create a smaller binary size at the end and is recommended by pybind11, but with that option you need to explicitly make a function (like the module init function) global.

File: docs/source/installing.rst [hunk 3]
```diff
@@ -49,7 +49,8 @@ If you want to work on vaex for a Pull Request from the source, use the followin
 * make sure the dev versions of pcre are installed (e.g. ``conda install -c conda-forge pcre``)
 * install using (note: if you're on Windows, make sure that your command line/terminal has administrator privileges):
 
- * ``pip install -e .``  (again, use (ana)conda or virtualenv/venv)
+ * Create and activate a virtualenv/venv
+ * ``make init``
```
[the remark this answers is not in the document]
Comment (Alon-Alexander) [reply docs/source/installing.rst:53]: If make doesn't work on windows, they won't be able to use the makefile at all :)
Do you have another good alternative? Because I don't want to also inline the `make deps`, that would be messy.

[the remark this answers is not in the document]
Comment (Alon-Alexander) [reply docs/source/installing.rst:53]: Someone without `make` won't be able to use that file at all, and that's fine because it's a utility file.
But if they use the makefile they probably have `make`.
Do you have another solution for that? Because I don't want to change `make clean` for example, as it would be very messy.

File: packages/vaex-core/src/vaexfast.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (maartenbreddels) [reply packages/vaex-core/src/vaexfast.cpp:2125]:
```suggestion
```
compiler doesn't seem to like it.

File: packages/vaex-core/setup.py [hunk 4]
```diff
@@ -75,14 +60,12 @@ def __str__(self):
     extra_compile_args.append("-mmacosx-version-min=10.9")
 
 # on windows (Conda-forge builds), the dirname is an absolute path
-extension_vaexfast = Extension("vaex.vaexfast", [os.path.relpath(os.path.join(dirname, "src/vaexfast.cpp"))],
+extension_vaexfast = Pybind11Extension("vaex.vaexfast", [os.path.relpath(os.path.join(dirname, "src/vaexfast.cpp"))],
```
Comment (maartenbreddels) [inline packages/vaex-core/setup.py:63]: Not using Pybind11 here:
```suggestion
extension_vaexfast = Extension("vaex.vaexfast", [os.path.relpath(os.path.join(dirname, "src/vaexfast.cpp"))],
```

File: packages/vaex-core/setup.py [hunk 5]
```diff
@@ -2,7 +2,7 @@
 import sys
 import os
 import imp
-from setuptools import Extension
+from pybind11.setup_helpers import Pybind11Extension
```
Comment (maartenbreddels) [inline packages/vaex-core/setup.py:5]: still need the import 
```suggestion
from setuptools import Extension
from pybind11.setup_helpers import Pybind11Extension
```

**Comment by maartenbreddels** [comment]
I wonder if pybind11 adds some compilation flags, because now the vaexfast extension doesn't export a symbol anymore 🤷

**Comment by Alon-Alexander** [comment]
I see your point @maartenbreddels 
I do know that pybind adds a few flags (this is why I added the attribute), but as it looks it changes something globally (as it doesn't expose the symbol in the changes you made).
Why don't we compile everything with pybind? Do we care?
Another solution would be to add the attribute only on non-windows (as I should have originally added it 😅 ).

**Comment by maartenbreddels** [comment]
> Why don't we compile everything with pybind?

Yeah, I'm good with that, but CI failed with it, maybe we should also use `PYBIND11_MODULE` in vaexfast.cpp, do you want to try that?

**Comment by maartenbreddels** [comment]
I don't know why the linux build has this visibility issue, lost there 🤷

**Comment by Alon-Alexander** [comment]
I noticed that most changes we talked about in this PR are unrelated to the change I wanted to make and belong in another refactoring PR.
Since I really want to merge the changes that will help other people develop vaex, I reverted most changes - the ones that refactored old code instead of addressing the issue I wanted to address.

File: docs/Makefile [hunk 6]
```diff
@@ -3,7 +3,7 @@
 
 # You can set these variables from the command line.
 SPHINXOPTS    =
-SPHINXBUILD   = sphinx-build
+SPHINXBUILD   = ../venv/bin/sphinx-build
```
Comment (maartenbreddels) [inline docs/Makefile:6]: I'll revert this since it's very specific to use of venv
```suggestion
SPHINXBUILD   = sphinx-build
```

File: docs/source/conf.py [hunk 7]
```diff
@@ -174,7 +174,7 @@ def system(cmd):
 extra_navbar = "<h1>dsadsa</h2>"
 
 html_sidebars = {
-	"**": ["sbt-sidebar-footer.html", "sbt-sidebar-nav.html", "sidebar-search-bs.html"]
+	"**": ["sbt-sidebar-footer.html", "sbt-sidebar-nav.html", "searchbox.html"]
```
Comment (maartenbreddels) [inline docs/source/conf.py:177]: why was this needed?

File: docs/source/installing.rst [hunk 8]
```diff
@@ -44,7 +44,7 @@ If you want to work on vaex for a Pull Request from the source, use the followin
 * make sure the dev versions of pcre are installed (e.g. ``conda install -c conda-forge pcre``)
 * install using (note: if you're on Windows, make sure that your command line/terminal has administrator privileges):
 
- * ``pip install -e .``  (again, use (ana)conda or virtualenv/venv)
+ * ``make init`` or ``pip install -e .[dev]``  (again, use (ana)conda or virtualenv/venv)
```
Comment (maartenbreddels) [inline docs/source/installing.rst:47]: zsh friendly
```suggestion
 * ``make init`` or ``pip install -e ".[dev]"``  (again, use (ana)conda or virtualenv/venv)
```

**Review by maartenbreddels — COMMENTED** [review]
yeah, good idea, really want to get this in

**Comment by maartenbreddels** [comment]
old
image
vs new:
image
I think I prefer the old one

File: docs/source/conf.py (hunk 7 above)
Comment (maartenbreddels) [reply docs/source/conf.py:177]:
```suggestion
	"**": ["sbt-sidebar-footer.html", "sbt-sidebar-nav.html", "sidebar-search-bs.html"]
```
lets revert (see my comment)

**Comment by Alon-Alexander** [comment]
What are we waiting for?
Is there anything else which requires change?

**Comment by maartenbreddels** [comment]
No, was waiting for CI, failures are unrelated (tensorflow version issues, if you want to take a look at that, feel free).
Will merge, thanks a lot!
